ddl/placement/constraint_test.go

```diff
@@ -0,0 +1,292 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package placement
+
+import (
+	"errors"
+
+	. "github.com/pingcap/check"
+)
+
+var _ = Suite(&testConstraintSuite{})
+
+type testConstraintSuite struct{}
+
+func (t *testConstraintSuite) TestNew(c *C) {
+	type TestCase struct {
+		name  string
+		input string
+		label Constraint
+		err   error
+	}
+	tests := []TestCase{
+		{
+			name:  "normal",
+			input: "+zone=bj",
+			label: Constraint{
+				Key:    "zone",
+				Op:     In,
+				Values: []string{"bj"},
+			},
+		},
+		{
+			name:  "normal with spaces",
+			input: "-  dc  =  sh  ",
+			label: Constraint{
+				Key:    "dc",
+				Op:     NotIn,
+				Values: []string{"sh"},
+			},
+		},
+		{
+			name:  "not tiflash",
+			input: "-engine  =  tiflash  ",
+			label: Constraint{
+				Key:    "engine",
+				Op:     NotIn,
+				Values: []string{"tiflash"},
+			},
```

Comment:
Should it have some tests where the case of `tiflash` is upper? The code lowers it for comparison, but does not save it in values as lower as it parses it (making it compatible with previous code). But I'm wondering if it should lower when saving in values so the string is effectively normalized (space removed, consistent case etc.)

Reply:
I think there is one test for it, the typo pointed out by you.

```
	{
			name:  "disallow tiflash",
			input: "+engine=Tiflash",
			err:   ErrUnsupportedConstraint,
		},
```

> But I'm wondering if it should lower when saving in values so the string is effectively normalized.

There should not be any normalization during comparison. I mean, I only do lowering for TiFlash, and it is a special case. It seems that we have no official doc, or spec for upper/lower cases of engine labels, so i added lower checking for a more relax check(so that tiflash will be banned even if it is TiFlash, Tiflash, or whatever).

Reply:
It is a bit of a strange behavior if the token TiFlash is case insensitive, but other tokens are stored in original case?

I think tokens are always case insensitive in MySQL, but in either case you decide (or don't decide), I think it makes sense to capture the current expected behavior in tests:

```
input: "+ZONE=BJ",
input: "+Zone=Bj",
```

Reply:
The thing is it is not a token, but a string. The syntax is like `CONSTRIANTS="+zone=bj"`. Here we use this string a "pattern matcher", which will match constraints added by user, or the software itself. PD and TiDB both have a matcher: TiDB will match on the constraints cache, PD will match on the persisted constraints.

So, there are other two concerns:
1. `zone,bj` kv pair is added by a user. But possibly not by a SQL of TiDB, e.g. tidb-operator could add such label. They won't normalize the label.
2. PD match case insensitively. It is also strange that TiDB is insensitive, but PD is sensitive.

For `TiFlash`, it is just, I don't know, if the value is always `TiFlash`. I mean, there is no spec for that. And this label is absolutely added by TiKV,TiFlash,TiDB. User should not add `engine=xxx` label. It will break the system.

Reply:
The thing is it is not a token, but a string. The syntax is like `CONSTRIANTS="+zone=bj"`. Here we use this string a "pattern matcher", which will match constraints added by user, or the software itself. PD and TiDB both have a matcher: TiDB will match on the constraints cache, PD will match on the persisted constraints.

So, there are other two concerns:
1. `zone,bj` kv pair is added by a user. But possibly not by a SQL of TiDB, e.g. tidb-operator could add such label. They won't normalize the label.
2. PD match case insensitively. It is also strange that TiDB is insensitive, but PD is sensitive.

For `TiFlash`, it is just, I don't know, if the value is always `TiFlash`. I mean, there is no spec for that. And this label is absolutely added by TiKV,TiFlash,TiDB. User should not add `engine=xxx` label. It will break the system.

Reply:
OK. It seems complicated enough maybe we can't solve it in this PR. LGTM.